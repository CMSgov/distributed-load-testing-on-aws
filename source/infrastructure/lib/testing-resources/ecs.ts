// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Aws, CfnResource, Tags, Fn } from "aws-cdk-lib";
import { ContainerImage, FargateTaskDefinition, LogDriver, CfnCluster } from "aws-cdk-lib/aws-ecs";
import { Effect, ServicePrincipal, Role, PolicyDocument, PolicyStatement } from "aws-cdk-lib/aws-iam";
import { LogGroup, RetentionDays } from "aws-cdk-lib/aws-logs";
import { CfnSecurityGroup, CfnSecurityGroupEgress, CfnSecurityGroupIngress } from "aws-cdk-lib/aws-ec2";
import { Bucket } from "aws-cdk-lib/aws-s3";
import { Construct } from "constructs";
import { DockerImageAsset } from "aws-cdk-lib/aws-ecr-assets";
import * as path from "path";
import { addCfnGuardSuppression } from "../common-resources/add-cfn-guard-suppression";

export interface ECSResourcesConstructProps {
  readonly fargateVpcId: string;
  readonly scenariosS3Bucket: string;
  readonly securityGroupEgress: string;
  readonly solutionId: string;
  readonly stableTagCondition: string;
}

/**
 * Distributed Load Testing on AWS Fargate and ECS test runner construct.
 * This creates the ECS cluster, Fargate task definition, and Security Group
 */
export class ECSResourcesConstruct extends Construct {
  public taskClusterName: string;
  public ecsCloudWatchLogGroup: LogGroup;
  public taskDefinitionArn: string;
  public taskExecutionRoleArn: string;
  public ecsSecurityGroupId: string;
  public taskDefinitionFamily: string;

  constructor(scope: Construct, id: string, props: ECSResourcesConstructProps) {
    super(scope, id);

    const dltTaskCluster = new CfnCluster(this, "DLTEcsCluster", {
      clusterName: Aws.STACK_NAME,
      clusterSettings: [{ name: "containerInsights", value: "enabled" }],
      tags: [
        {
          key: "SolutionId",
          value: props.solutionId,
        },
        {
          key: "CloudFormation Stack",
          value: Aws.STACK_NAME,
        },
      ],
    });

    this.taskClusterName = dltTaskCluster.ref;

    const scenariosBucketArn = Bucket.fromBucketName(this, "ScenariosBucket", props.scenariosS3Bucket).bucketArn;

    const dltTaskExecutionRole = new Role(this, "DLTTaskExecutionRole", {
      assumedBy: new ServicePrincipal("ecs-tasks.amazonaws.com"),
      inlinePolicies: {
        ECSTaskExecutionPolicy: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: [
                "ecr:GetAuthorizationToken",
                "ecr:BatchCheckLayerAvailability",
                "ecr:GetDownloadUrlForLayer",
                "ecr:BatchGetImage",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              resources: ["*"],
            }),
          ],
        }),
        ScenariosS3Policy: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["s3:HeadObject", "s3:PutObject", "s3:GetObject", "s3:ListBucket"],
              resources: [scenariosBucketArn, `${scenariosBucketArn}/*`],
            }),
          ],
        }),
      },
    });
    this.taskExecutionRoleArn = dltTaskExecutionRole.roleArn;

    addCfnGuardSuppression(dltTaskExecutionRole, "IAM_POLICYDOCUMENT_NO_WILDCARD_RESOURCE");
    addCfnGuardSuppression(dltTaskExecutionRole, "IAM_NO_INLINE_POLICY_CHECK");

    this.ecsCloudWatchLogGroup = new LogGroup(this, "DLTCloudWatchLogsGroup", {
      retention: RetentionDays.ONE_YEAR,
    });
    const dltLogsGroupResource = this.ecsCloudWatchLogGroup.node.defaultChild as CfnResource;
    dltLogsGroupResource.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W84",
          reason: "KMS encryption unnecessary for log group",
        },
      ],
    });
    Tags.of(this.ecsCloudWatchLogGroup).add("SolutionId", props.solutionId);

    const dltTaskDefinition = new FargateTaskDefinition(this, "DLTTaskDefinition", {
      cpu: 2048,
      memoryLimitMiB: 4096,
      executionRole: dltTaskExecutionRole,
      taskRole: dltTaskExecutionRole,
    });
    const dockerRepoName = "distributed-load-testing-on-aws-load-tester";

    const versionTagForImage: string =
      process.env.PUBLIC_ECR_REGISTRY && process.env.PUBLIC_ECR_TAG
        ? `${process.env.PUBLIC_ECR_REGISTRY}/${dockerRepoName}:${process.env.PUBLIC_ECR_TAG}`
        : "";

    const stageTagForImage: string =
      process.env.PUBLIC_ECR_REGISTRY && process.env.PUBLIC_ECR_TAG
        ? `${process.env.PUBLIC_ECR_REGISTRY}/${dockerRepoName}:${
            process.env.PUBLIC_ECR_TAG.substring(0, 4) + "_stable"
          }`
        : "";

    const imageUseTag = Fn.conditionIf(props.stableTagCondition, stageTagForImage, versionTagForImage).toString();

    const imageAsset = imageUseTag
      ? ContainerImage.fromRegistry(`${imageUseTag}`)
      : new DockerImageAsset(this, "LoadTesterImage", {
          directory: path.join(__dirname, `../../../../deployment/ecr/${dockerRepoName}`),
        });

    dltTaskDefinition.addContainer("LoadTestContainer", {
      containerName: `${Aws.STACK_NAME}-load-tester`,
      image: imageAsset instanceof DockerImageAsset ? ContainerImage.fromDockerImageAsset(imageAsset) : imageAsset,
      memoryLimitMiB: 4096,
      logging: LogDriver.awsLogs({
        streamPrefix: "load-testing",
        logGroup: this.ecsCloudWatchLogGroup,
      }),
    });
    Tags.of(dltTaskDefinition).add("SolutionId", props.solutionId);
    this.taskDefinitionArn = dltTaskDefinition.taskDefinitionArn;

    const ecsSecurityGroup = new CfnSecurityGroup(this, "DLTEcsSecurityGroup", {
      vpcId: props.fargateVpcId,
      groupDescription: "DLTS Tasks Security Group",
    });
    ecsSecurityGroup.addMetadata("cfn_nag", {
      rules_to_suppress: [
        {
          id: "W40",
          reason: "IpProtocol set to -1 (any) as ports are not known prior to running tests",
        },
        {
          id: "F1000",
          reason: "egress rule is specificed as its own cfn resource.",
        },
      ],
    });

    this.ecsSecurityGroupId = ecsSecurityGroup.ref;

    new CfnSecurityGroupEgress(this, "DLTSecGroupEgress", {
      cidrIp: props.securityGroupEgress,
      description: "Allow tasks to call out to external resources",
      groupId: ecsSecurityGroup.ref,
      ipProtocol: "-1",
    });

    new CfnSecurityGroupIngress(this, "DLTSecGroupIngress", {
      description: "Allow tasks to communicate",
      fromPort: 50000,
      groupId: ecsSecurityGroup.ref,
      ipProtocol: "tcp",
      sourceSecurityGroupId: ecsSecurityGroup.ref,
      toPort: 50000,
    });

    this.taskDefinitionFamily = dltTaskDefinition.family;
  }
}
