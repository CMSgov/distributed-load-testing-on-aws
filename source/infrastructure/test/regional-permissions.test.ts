// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import '@aws-cdk/assert/jest';
import { SynthUtils } from '@aws-cdk/assert';
import { Stack } from 'aws-cdk-lib';
import { RegionalPermissionsConstruct } from '../lib/testing-resources/regional-permissions';

test('DLT Regional Permission Test', () => {
  const stack = new Stack();

  new RegionalPermissionsConstruct(stack, 'TestRegionalPermissions', {
    apiServicesLambdaRoleName: 'testApiServicesLambdaRoleName',
    resultsParserRoleName: 'testResultsParserRoleName',
    taskExecutionRoleArn: 'arn:aws:iam::123456789012:role/testRole',
    ecsCloudWatchLogGroupArn: 'arn:aws:logs:us-east-2:123456789012:log-group:test_log_group_name',
    taskRunnerRoleName: 'testTaskRunnerRoleName',
    taskCancelerRoleName: 'testTaskCancelerRoleName',
    taskStatusCheckerRoleName: 'testTaskStatusCheckerRoleName'
  });
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  expect(stack).toHaveResourceLike("AWS::IAM::Policy", {
    PolicyDocument: {
      Statement: [
        {
          Action: [
            "ecs:RunTask",
            "ecs:DescribeTasks",
          ],
          Effect: "Allow",
          Resource: [
            {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    Ref: "AWS::Partition",
                  },
                  ":ecs:",
                  {
                    Ref: "AWS::Region",
                  },
                  ":",
                  {
                    Ref: "AWS::AccountId",
                  },
                  ":task/*",
                ],
              ],
            },
            {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    Ref: "AWS::Partition",
                  },
                  ":ecs:",
                  {
                    Ref: "AWS::Region",
                  },
                  ":",
                  {
                    Ref: "AWS::AccountId",
                  },
                  ":task-definition/*:*",
                ],
              ],
            },
          ],
        },
        {
          Action: "iam:PassRole",
          Effect: "Allow",
          Resource: "arn:aws:iam::123456789012:role/testRole",
        },
      ],
      Version: "2012-10-17",
    },
    PolicyName: {
      "Fn::Join": [
        "",
        [
          "RegionalECRPerms-",
          {
            Ref: "AWS::StackName",
          },
          "-",
          {
            Ref: "AWS::Region",
          }
        ]
      ]
    }
  });
  expect(stack).toHaveResourceLike("AWS::IAM::Policy", {
    PolicyDocument: {
      Statement: [
        {
          Action: "logs:PutMetricFilter",
          Effect: "Allow",
          Resource: "arn:aws:logs:us-east-2:123456789012:log-group:test_log_group_name",
        },
      ],
      Version: "2012-10-17",
    }
  });
  expect(stack).toHaveResourceLike("AWS::IAM::Policy", {
    PolicyDocument: {
      Statement: [
        {
          Action: "logs:DeleteMetricFilter",
          Effect: "Allow",
          Resource: "arn:aws:logs:us-east-2:123456789012:log-group:test_log_group_name",
        },
      ],
      Version: "2012-10-17",
    }
  });
  expect(stack).toHaveResourceLike("AWS::IAM::Policy", {
    PolicyDocument: {
      Statement: [
        {
          Action: "ecs:StopTask",
          Effect: "Allow",
          Resource: [
            {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    Ref: "AWS::Partition",
                  },
                  ":ecs:",
                  {
                    Ref: "AWS::Region",
                  },
                  ":",
                  {
                    Ref: "AWS::AccountId",
                  },
                  ":task/*",
                ],
              ],
            },
            {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    Ref: "AWS::Partition",
                  },
                  ":ecs:",
                  {
                    Ref: "AWS::Region",
                  },
                  ":",
                  {
                    Ref: "AWS::AccountId",
                  },
                  ":task-definition/*:*",
                ],
              ],
            },
          ],
        },
      ],
      Version: "2012-10-17",
    }
  });
  expect(stack).toHaveResourceLike("AWS::IAM::Policy", {
    PolicyDocument: {
      Statement: [
        {
          Action: "ecs:DescribeTasks",
          Effect: "Allow",
          Resource: {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  Ref: "AWS::Partition",
                },
                ":ecs:",
                {
                  Ref: "AWS::Region",
                },
                ":",
                {
                  Ref: "AWS::AccountId",
                },
                ":task/*",
              ],
            ],
          },
        },
      ],
      Version: "2012-10-17",
    }
  });
});