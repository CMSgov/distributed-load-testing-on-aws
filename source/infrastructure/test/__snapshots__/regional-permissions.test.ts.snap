// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DLT Regional Permission Test 1`] = `
Object {
  "Resources": Object {
    "TestRegionalPermissionsECSCloudWatchDelMetrics281D272A": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "logs:DeleteMetricFilter",
              "Effect": "Allow",
              "Resource": "arn:aws:logs:us-east-2:123456789012:log-group:test_log_group_name",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "ECSCloudWatchDelMetrics-",
              Object {
                "Ref": "AWS::StackName",
              },
              "-",
              Object {
                "Ref": "AWS::Region",
              },
            ],
          ],
        },
        "Roles": Array [
          "testApiServicesLambdaRoleName",
          "testResultsParserRoleName",
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestRegionalPermissionsECSCloudWatchPutMetricsdD0A5AD2E": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "logs:PutMetricFilter",
              "Effect": "Allow",
              "Resource": "arn:aws:logs:us-east-2:123456789012:log-group:test_log_group_name",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "ECSCloudWatchPutMetrics-",
              Object {
                "Ref": "AWS::StackName",
              },
              "-",
              Object {
                "Ref": "AWS::Region",
              },
            ],
          ],
        },
        "Roles": Array [
          "testTaskRunnerRoleName",
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestRegionalPermissionsECSDescribePolicyAA323A68": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "ecs:DescribeTasks",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":ecs:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":task/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "ECSDescribePolicy",
              Object {
                "Ref": "AWS::Region",
              },
            ],
          ],
        },
        "Roles": Array [
          "testTaskStatusCheckerRoleName",
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestRegionalPermissionsECSStopPolicy4B5507E2": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "ecs:StopTask",
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":ecs:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":task/*",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":ecs:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":task-definition/*:*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "ECSStopPolicy-",
              Object {
                "Ref": "AWS::StackName",
              },
              "-",
              Object {
                "Ref": "AWS::Region",
              },
            ],
          ],
        },
        "Roles": Array [
          "testTaskCancelerRoleName",
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestRegionalPermissionsRegionalECRPermsDD807682": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "ecs:RunTask",
                "ecs:DescribeTasks",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":ecs:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":task/*",
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":ecs:",
                      Object {
                        "Ref": "AWS::Region",
                      },
                      ":",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":task-definition/*:*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "iam:PassRole",
              "Effect": "Allow",
              "Resource": "arn:aws:iam::123456789012:role/testRole",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "RegionalECRPerms-",
              Object {
                "Ref": "AWS::StackName",
              },
              "-",
              Object {
                "Ref": "AWS::Region",
              },
            ],
          ],
        },
        "Roles": Array [
          "testTaskRunnerRoleName",
          "testApiServicesLambdaRoleName",
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
